### ISSUE TITLE
Multiple issues when installing on Ubuntu 24.04

### ISSUE THREAD

**Issue by pgrenaud** [six2dez/reconftw#907]
### TL;DR

- Single apt package with multiple candidates prevent installation of all apt packages
- Python new _Externally Managed Environments_ feature prevents installation of all tools that rely on pip
- Muted output during installation prevents troubleshooting

### Description

I recently tried to install reconftw on a Ubuntu 24.04 and notice that something was wrong. The `install.sh` script would report that everything was installed while the `reconftw.sh` script would report a bunch of tools missing. Running `install.sh` again wouldn't fix the issue.

![Screenshot 2024-11-06 at 14 06 49](https://github.com/user-attachments/assets/e3902ac4-e12c-4aa5-827b-fcb36de163dc)
![Screenshot 2024-11-06 at 14 06 57](https://github.com/user-attachments/assets/07d8a849-da55-4b33-b42d-39fafedb0bc9)

Re-enabling output during the install (vars `DEBUG_STD` and `DEBUG_ERROR` in `reconftw.cfg`) allowed me to notice a few problems with the install process.

The installing system packages step failed silently because the package `libasound2` is a virtual package with two possible candidates (this is unique to Ubuntu 24.04). The resulting error blocks the installation of all required system packages.

![Screenshot 2024-11-06 at 14 11 05](https://github.com/user-attachments/assets/73067557-bc8e-427e-ab72-8b88ecbc3a97)

Many python tools rely on PIP to install their requirements, but reconftw do it by running `sudo pip3 install -r requirements.txt`, which is now blocked by the _Externally Managed Environments_ feature.

![Screenshot 2024-11-06 at 14 14 01](https://github.com/user-attachments/assets/ed5fe072-1e06-4323-96a1-a4b8462ec839)

There is also a bunch more errors that are probably caused by the missing apt packages and/or the missing python dependencies.

### Affected Systems

- Ubuntu 24.04 (Noble)
- Debian 12 (Bookworm)
- Kali Linux (exact versions unknown) (see #899)

### Expected Results

Cloning the repo and running `./install.sh` should install all the required dependencies and tools.

### Actual Results

Cloning the repo and running `./install.sh` is silently falling and not installing all the dependencies and tools. See screenshots above.

### Steps to Reproduce

- Deploy a new Ubuntu 24.04 LTS machine
- Clone and install reconftw
  ```bash
  git clone https://github.com/six2dez/reconftw.git
  cd reconftw/
  ./install.sh
  ```
- Run `./reconftw.sh --help`
- reconftw will report several tools missing

### Discussion and Proposed Solutions

I've been using reconftw for the past year at work and it has been a great tool to have in my toolbox! I'm ready to put the time and effort to fix the issues highlighted and submit PRs for review. But since there is no contributing guide for this project, I just want to make sure we all agree on the work that need to be done before starting to work on fixes.

I'd like to propose the following changes:

1. Send all the output of the `./install.sh` to a log file by default, like it is already the case with `./reconftw.sh`
2. Fix the list of apt packages to install (not sure yet how to do it in a compatible manner with all the debian-based distributions, would need to look into it)
3. Install all python tools using `pipx`, or setup a `venv` for those who do not support it
4. Remove the use of `sudo` when it's not necessary (all tools would be installed for the local user only, like it is the case with the golang tools)
5. In the long term, validate the exit code of commands, especially for`./install.sh`, to make the scripts more resilient and detect abnormal states and errors

I can break down this issue in multiple issues if that would make it easier to keep track of the different issues and proposed changes.

### See Also

- https://packaging.python.org/en/latest/specifications/externally-managed-environments/#guide-users-towards-virtual-environments
- https://peps.python.org/pep-0668/
- https://www.kali.org/blog/python-externally-managed/

**Comment by bileltechno** [comment]
Thanks for your detailed  report! 

Some tools installation is already added to the auto installer in the dev branch but not yet merged to the main branch.


[adding EmailFinder to auto isntaller](https://github.com/six2dez/reconftw/commit/87b28d02f259b31b39d53a46ee4abe96bf9656dd)

[add metaFinder to auto installer](https://github.com/six2dez/reconftw/commit/fc608c487122ef4a3eb84af7af43c9fce0fd4f81)

[add porch-pirate to auto-install](https://github.com/six2dez/reconftw/commit/c564fdc966907feb611f0a42b2cd7c172e44f759)

[Add xnLinkFinder to auto installer](https://github.com/six2dez/reconftw/commit/ed6917ce619740982f148213b2986180bf9b93e4)

[add waymore to auto-installer](https://github.com/six2dez/reconftw/commit/688e5d6de007d4ba30f081b92c69d7eeb40cbfd3)

The remaining issues will be analysed later .

**Comment by pgrenaud** [comment]
> Some tools installation is already added to the auto installer in the dev branch but not yet merged to the main branch.

I don't think those commits address the problems highlighted in this issues. Unless I missed something, but when running on Ubuntu 22.04 (or Ubuntu 24.04 with some hotfixes to make the `install.sh` script work), everything is installed and the main `reconftw.sh` script works properly.

> The remaining issues will be analysed later .

Ok, no problem! Please let me know what you think of my diagnostic and proposed solutions. I'm able to put effort into writing and sending PR to fix those issues.

**Comment by g1thubcrypton** [comment]
i have the similar problem

**Comment by ab2pentest** [comment]
The problem is with **Python 3.12.X**, which now requires the `--break-system-packages` flag to install packages on the system, unless you create a separate environment for each tool.

If you're using a virtual machine, VPS, or similar setup just for pentesting, you can fix this by editing the install.sh file. Find this line:

```
eval "$SUDO pip3 install . $DEBUG_STD" &>/dev/null
```

Change it to this:

```
eval "$SUDO pip3 install . --break-system-packages $DEBUG_STD" &>/dev/null
```

Then, run the install.sh script again to install everything.

**Comment by six2dez** [comment]
switched to pipx for installing requirements and tools, could you check it?

**Comment by pgrenaud** [comment]
@six2dez Sorry for the delay! I finally got time to look at it this week at work.

I looked at both 56a4ce78fee196268c0643f32fe6d26497c15122 and 7f8717fee0af5416e3131bb620177d383e306f71. Unfortunately, the first commit does not work (that is not how packages are installed with `pipx`) and second commit does nothing (removing `&>/dev/null` reveals errors that were hidden).

So, instead of going back and forth to fix this, I worked in the past two days on PRs to address some of the issues I've notice. Unluckily, I uncovered a lot of issues while working on the PRs, which significantly increased the amount of time I initially thought it would take me to do the work.

Two notables things I want to mention:
- One of my first commit was to remove all `&>/dev/null` from the `install.sh` script, which made debugging lot easier. I won't push this commit, but I'm really eager to know the reason why all the output is suppressed instead of being redirected like it was before. I would really like to find a better solution to this that would satisfy all of us.
- An other commit I made earlier was to add a GitHub workflow to test the `install.sh` script on various platforms. That really help me throughout the integration of `pipx`, but I also uncovered a lot of hidden/unknown issues by doing so. I will push this PR first, which will definitely help detecting broken features and regressions in the future.

I'll break the changes into multiple smaller PRs and try to provide context for each of them. However, you might have to deal with one or two merge conflicts when merging everything back together. **If you do prefer one single PR, let me know, I have a branch ready for that as well.**

**Comment by pgrenaud** [comment]
@six2dez I made a 6th PR to restore changes that were lost in the merges.

**Comment by pgrenaud** [comment]
Thanks for merging everything! I'll open a new issue for the other changes I'd like to discuss.
